File: 099_LCAInBST.cpp
/*
Problem: LCA in BST
Description: Find lowest common ancestor in Binary Search Tree.

Approach: Use BST property to navigate - if both nodes on same side, go that way.
*/

#include <bits/stdc++.h>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left, *right;
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
};

TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
    if (!root) return nullptr;
    
    if (p->val < root->val && q->val < root->val) {
        return lowestCommonAncestor(root->left, p, q);
    }
    
    if (p->val > root->val && q->val > root->val) {
        return lowestCommonAncestor(root->right, p, q);
    }
    
    return root;
}

int main() {
    TreeNode* root = new TreeNode(6);
    root->left = new TreeNode(2);
    root->right = new TreeNode(8);
    root->left->left = new TreeNode(0);
    root->left->right = new TreeNode(4);
    root->right->left = new TreeNode(7);
    root->right->right = new TreeNode(9);
    
    TreeNode* lca = lowestCommonAncestor(root, root->left, root->right);
    cout << lca->val << endl;  // Output: 6
    
    return 0;
}

/*
Time Complexity: O(h)
Space Complexity: O(h)
*/

File: 099_LCAInBST.py
"""
Problem: LCA in BST
Description: Find lowest common ancestor in Binary Search Tree.

Approach: Use BST property to navigate - if both nodes on same side, go that way.
"""

from typing import Optional

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

def lowestCommonAncestor(root: Optional[TreeNode], p: TreeNode, q: TreeNode) -> Optional[TreeNode]:
    if not root:
        return None
    
    if p.val < root.val and q.val < root.val:
        return lowestCommonAncestor(root.left, p, q)
    
    if p.val > root.val and q.val > root.val:
        return lowestCommonAncestor(root.right, p, q)
    
    return root

def main():
    root = TreeNode(6)
    root.left = TreeNode(2)
    root.right = TreeNode(8)
    root.left.left = TreeNode(0)
    root.left.right = TreeNode(4)
    root.right.left = TreeNode(7)
    root.right.right = TreeNode(9)
    
    lca = lowestCommonAncestor(root, root.left, root.right)
    print(lca.val)  # Output: 6

if __name__ == "__main__":
    main()

"""
Time Complexity: O(h)
Space Complexity: O(h)
"""
